workspaces/optic-engine/src/projections/learners/undocumented_bodies.rs

```diff
@@ -10,29 +10,69 @@ use crate::JsonTrail;
 
 #[derive(Default, Debug)]
 pub struct LearnedUndocumentedBodiesProjection {
-  observations_by_body_location: HashMap<BodyAnalysisLocation, TrailObservationsResult>,
+  request_response_observations: HashMap<BodyAnalysisLocation, TrailObservationsResult>,
+  query_params_observations: HashMap<BodyAnalysisLocation, TrailObservationsResult>,
 }
 
 impl LearnedUndocumentedBodiesProjection {
   fn with_body_analysis_result(&mut self, analysis: BodyAnalysisResult) {
-    let existing_observations = self
-      .observations_by_body_location
+    let collection = match analysis.body_location {
+      BodyAnalysisLocation::UnmatchedRequest { .. }
+      | BodyAnalysisLocation::UnmatchedResponse { .. } => &mut self.request_response_observations,
+      BodyAnalysisLocation::UnmatchedRequestQueryParameters { .. } => {
+        &mut self.query_params_observations
+      }
+      BodyAnalysisLocation::MatchedRequest { .. }
+      | BodyAnalysisLocation::MatchedResponse { .. } => {
+        unreachable!("analyzing undocumented bodies should only yield unmatched analysiss")
+      }
+    };
+
+    let existing_observations = collection
       .entry(analysis.body_location)
       .or_insert_with(|| TrailObservationsResult::default());
 
     existing_observations.union(analysis.trail_observations.normalized());
   }
 
   pub fn into_endpoint_bodies(
-    self,
+    mut self,
     id_generator: &mut impl SpecIdGenerator,
   ) -> impl Iterator<Item = EndpointBodies> {
     let mut endpoints_by_endpoint = HashMap::new();
-    for (body_location, observations) in self.observations_by_body_location {
+    for (body_location, observations) in self.request_response_observations {
       let (root_shape_id, body_commands) =
         observations.into_commands(id_generator, &JsonTrail::empty());
-      let endpoint_body =
-        EndpointBody::new(&body_location, root_shape_id, body_commands, id_generator);
+      let mut endpoint_body = EndpointBody::new(&body_location, root_shape_id, body_commands);
+
+      if let BodyAnalysisLocation::UnmatchedRequest {
+        ref path_id,
+        ref method,
+        ref content_type,
+      } = body_location
+      {
+        let expected_query_params_location =
+          BodyAnalysisLocation::UnmatchedRequestQueryParameters {
+            path_id: path_id.clone(),
+            method: method.clone(),
+            content_type: content_type.clone(),
```

Comment:
I've seen usages of `clone()` for `&str` through our code. Wouldn't it be better to pass the references through (and make sure the lifetimes we are using are correct)? Or are we doing this to avoid specifying lifetimes + so we can get something functional out faster, or do we expect to mutate these in each usage (or something else that I'm missing)

Reply:
In this case we need to clone, as we like the output type to own its data. That's the  most straightforward way to know it can be safely adapted, rewritten and serialised later on (even if JSON serde support serialising by reference). 

That said, it's also definitely because it's naive and less complicated way to do things at first, and nothing in the performance we've measured suggests it's worth the cost of optimising this (making it more complicated to write initially, but also change and maintain generally).